packages/mds-metrics-sheet/metrics-log-utils.ts

```diff
@@ -0,0 +1,191 @@
+  const d = new Date()
+  let [enters, leaves, starts, ends, start_sla, end_sla, telems, telem_sla] = [0, 0, 0, 0, 0, 0, 0, 0]
+  let event_counts = { service_start: 0, provider_drop_off: 0, trip_start: 0, trip_end: 0 }
+  let status_counts = {
+    available: 0,
+    unavailable: 0,
+    reserved: 0,
+    trip: 0,
+    removed: 0,
+    inactive: 0,
+    elsewhere: 0
+  }
+  const { event_counts_last_24h, late_event_counts_last_24h, late_telemetry_counts_last_24h } = last[
+    provider.provider_id
+  ]
+  if (event_counts_last_24h) {
+    event_counts = event_counts_last_24h
+    status_counts = eventCountsToStatusCounts(event_counts_last_24h)
+    starts = event_counts_last_24h.trip_start || 0
+    ends = event_counts_last_24h.trip_end || 0
+    enters = event_counts_last_24h.trip_enter || 0
+    leaves = event_counts_last_24h.trip_leave || 0
+    telems = last[provider.provider_id].telemetry_counts_last_24h || 0
+    if (late_telemetry_counts_last_24h !== undefined && late_telemetry_counts_last_24h !== null) {
+      telem_sla = telems ? percent(late_telemetry_counts_last_24h, telems) : 0
+    }
+    if (late_event_counts_last_24h !== undefined && late_event_counts_last_24h !== null) {
+      start_sla = starts ? percent(late_event_counts_last_24h.trip_start, starts) : 0
+      end_sla = ends ? percent(late_event_counts_last_24h.trip_end, ends) : 0
+    }
+  }
+  return {
+    date: `${d.toLocaleDateString('en-US', dateOptions)} ${d.toLocaleTimeString('en-US', timeOptions)}`,
+    name: provider.provider,
+    registered: provider.count || 0,
+    deployed:
+      sum([provider.status.available, provider.status.unavailable, provider.status.trip, provider.status.reserved]) ||
+      0,
+    validtrips: 'tbd', // Placeholder for next day valid trip analysis
+    trips: last[provider.provider_id].trips_last_24h || 0,
+    servicestart: event_counts.service_start || 0,
+    providerdropoff: event_counts.provider_drop_off || 0,
+    tripstart: starts,
+    tripend: ends,
+    tripenter: enters,
+    tripleave: leaves,
+    telemetry: telems,
+    telemetrysla: telem_sla,
+    tripstartsla: start_sla,
+    tripendsla: end_sla,
+    available: status_counts.available,
+    unavailable: status_counts.unavailable,
+    reserved: status_counts.reserved,
+    trip: status_counts.trip,
+    removed: status_counts.removed,
+    inactive: status_counts.inactive,
+    elsewhere: status_counts.elsewhere
+  }
+}
+
+export const getProviderMetrics = async (iter: number): Promise<ProviderMetrics> => {
+  const MAX_ITER = 10
+  /* after MAX_ITER failed iterations, give up */
+  if (iter >= MAX_ITER) {
+    throw new Error(`Failed to write to sheet after ${MAX_ITER} tries!`)
+  }
+  const token = await getAuthToken()
+  if (token == null) {
+    return getProviderMetrics(iter + 1)
+  }
+  const vehicleCounts = await getVehicleCounts(token)
+  const lastDayStats = await getLastDayStats(token)
```

Comment:
This `lastDayStats` is only used by one of the `getProviderMetrics()` consumers. I considered making this API call/result optional, but it adds some complexity and branching that IMO is not worth the potential efficiency gains.